HW2/P2/parallel_vector.pyx

```diff
@@ -107,9 +111,23 @@ cpdef move_data_medium_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0],num_threads=4): 
                if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+                   no_lock_src  = ( src[idx]+N-1)/N
+                   no_lock_dest = (dest[idx]+N-1)/N
+                   if no_lock_src == no_lock_dest :
```

Comment:
what if src[idx] or dest[idx] changes between now and acquiring the lock? then no_lock_src is invalidated. You usually want to check the same condition again after acquiring the lock